llvm/lib/Target/DirectX/DXIL.td

```diff
@@ -13,353 +13,777 @@
+def waveAllOp : DXILOpClass;
+def waveAllTrue : DXILOpClass;
+def waveAnyTrue : DXILOpClass;
+def waveGetLaneCount : DXILOpClass;
+def waveGetLaneIndex : DXILOpClass;
+def waveIsFirstLane : DXILOpClass;
+def waveMatch : DXILOpClass;
+def waveMatrix_Accumulate : DXILOpClass;
+def waveMatrix_Annotate : DXILOpClass;
+def waveMatrix_Depth : DXILOpClass;
+def waveMatrix_Fill : DXILOpClass;
+def waveMatrix_LoadGroupShared : DXILOpClass;
+def waveMatrix_LoadRawBuf : DXILOpClass;
+def waveMatrix_Multiply : DXILOpClass;
+def waveMatrix_ScalarOp : DXILOpClass;
+def waveMatrix_StoreGroupShared : DXILOpClass;
+def waveMatrix_StoreRawBuf : DXILOpClass;
+def waveMultiPrefixBitCount : DXILOpClass;
+def waveMultiPrefixOp : DXILOpClass;
+def wavePrefixOp : DXILOpClass;
+def waveReadLaneAt : DXILOpClass;
+def waveReadLaneFirst : DXILOpClass;
+def worldRayDirection : DXILOpClass;
+def worldRayOrigin : DXILOpClass;
+def worldToObject : DXILOpClass;
+def writeSamplerFeedback : DXILOpClass;
+def writeSamplerFeedbackBias : DXILOpClass;
+def writeSamplerFeedbackGrad : DXILOpClass;
+def writeSamplerFeedbackLevel : DXILOpClass;
+
+// This is a sentinel definition. Hence placed at the end here
+// and not as part of the above alphabetically sorted valid definitions.
+// It is never used to construct the name of DXIL Op call name.
+// Additionally it is capitalized unlike all the others.
+def UnknownOpClass : DXILOpClass;
+
+// Shader stages
+class ShaderStage;
+
+def compute : ShaderStage;
+def domain : ShaderStage;
+def hull : ShaderStage;
+def pixel : ShaderStage;
+def vertex : ShaderStage;
+def geometry : ShaderStage;
+def library : ShaderStage;
+def amplification : ShaderStage;
+def mesh : ShaderStage;
+def node : ShaderStage;
+def raygeneration : ShaderStage;
+def intersection : ShaderStage;
+def anyhit : ShaderStage;
+def closesthit : ShaderStage;
+def callable : ShaderStage;
+def miss : ShaderStage;
+
+// Primitive predicate
+class Pred;
+
+// Following classes represent predicate that evaluate to the payload,
+// if one xists, unconditionally.
+
+// DXIL Op attributes defined as true predicates, if specified
+
+def ReadOnly : Pred;
+def ReadNone : Pred;
+def IsDerivative : Pred;
+def IsGradient : Pred;
+def IsFeedback : Pred;
+def IsWave : Pred;
+def NeedsUniformInputs : Pred;
+def IsBarrier : Pred;
+
+// Predicate that evaluates to specified shader model version unconditionally.
+class SMVersion<Version ver> : Pred {
+  Version sm_version = ver;
+}
+
+// Predicate that evaluates to specified overload type list unconditionally.
+class Overloads<list<LLVMType> tys> : Pred {
+  list<LLVMType> overload_types = tys;
+}
+
+// Predicate that evaluates to specified list of shader stages unconditionally.
+class Stages<list<ShaderStage> st> : Pred {
+  list<ShaderStage> stage_kinds = st;
+}
+
+// Following class represents predicate that evaluates to the payload when
+// specified predicte is true.
+
+class Constraints<Pred p, list<Pred> l = []> : Pred {
+  Pred pred = p;
```

Comment:
It looks likke the tablegen code assumes that `pred` is always a `SMVersion`.  Could this be `SMVersion` instead?

Reply:
> It looks like the tablegen code assumes that `pred` is always a `SMVersion`. Could this be `SMVersion` instead?

Currently, the constraints are only based on Shader Model version. This class definition is intended to be generic enough to represent other constraints, if needed.